### PR TITLE
hls: avoid arithmetic between different enumeration types

### PR DESCRIPTION
**Pull request by smuzaffar**
- simplify hls patches
- Replace enum with `static constexpr` to avoid [warnings like](https://cmssdt.cern.ch/SDT/cgi-bin/logreader/el8_amd64_gcc12/CMSSW_14_1_X_2024-07-22-2300/pyRelValMatrixLogs/run/29634.702_TTbar_14TeV+2026D110_trackingMkFitPhase2/step3_TTbar_14TeV+2026D110_trackingMkFitPhase2.log#/)
```
In file included from DataFormatsL1TrackTrigger_xr dictionary payload:73:
In file included fromDataFormats/L1TrackTrigger/interface/TTTypes.h:19:
In file included from DataFormats/L1TrackTrigger/interface/TTTrack.h:18:
In file included from DataFormats/L1TrackTrigger/interface/TTTrack_TrackWord.h:20:
In file included from hls/2019.08-0e37f055a3ed22611ce5edecb14d0695/include/ap_int.h:20:
In file included from hls/2019.08-0e37f055a3ed22611ce5edecb14d0695/include/ap_common.h:252:
hls/2019.08-0e37f055a3ed22611ce5edecb14d0695/include/etc/ap_private.h:4550:58: warning: arithmetic between different enumeration types ('ap_private<128, false, false>::(unnamed enum at hls/2019.08-0e37f055a3ed22611ce5edecb14d0695/include/etc/ap_private.h:3025:3)' and 'ap_private<128, false, false>::(unnamed enum at hls/2019.08-0e37f055a3ed22611ce5edecb14d0695/include/etc/ap_private.h:3160:3)') is deprecated [-Wdeprecated-anon-enum-enum-conversion]
    if (this != &RHS) memcpy(pVal, RHS.get_pVal(), _AP_N * APINT_WORD_SIZE);

```

### CODE REVIEW

**Comment by smuzaffar** [comment]
please test workflow 29634.702

File: hls-fix.patch [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+     check_canary();
+@@ -3022,8 +3022,9 @@ class ap_private<_AP_W, _AP_S, false> {
+ #pragma warning(disable : 4521 4522)
+ #endif
+  public:
+-  enum { BitWidth = _AP_W, _AP_N = (_AP_W + 63) / 64 };
+-  static const int width = _AP_W;
++  static constexpr uint32_t BitWidth = _AP_W;
++  static constexpr uint32_t _AP_N = (_AP_W + 63) / 64;
++  constexpr static const int width = _AP_W;
```
Comment (smuzaffar) [inline hls-fix.patch]: @makortel , converting `enum` to `static constexpr` fixed the runtime cling warnings like https://github.com/cms-sw/cmsdist/pull/9311#issue-2424355867 . Does this change look reasonable to you?

File: hls-fix.patch [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+@@ -3157,10 +3158,8 @@ class ap_private<_AP_W, _AP_S, false> {
+   INLINE void set_pVal(int i, uint64_t value) { pVal[i] = value; }
+ 
+   /// This enum is used to hold the constants we needed for ap_private.
+-  enum {
+-    APINT_BITS_PER_WORD = sizeof(uint64_t) * 8, ///< Bits in a word
+-    APINT_WORD_SIZE = sizeof(uint64_t)          ///< Byte size of a word
+-  };
++  static constexpr int APINT_BITS_PER_WORD = sizeof(uint64_t) * 8;
++  static constexpr int APINT_WORD_SIZE = sizeof(uint64_t);
```
Comment (smuzaffar) [inline hls-fix.patch:48]: @makortel , same here `enum` to `static constexpr`

File: hls-fix.patch (hunk 1 above)
Comment (iarspider) [reply hls-fix.patch]: I think `const` on L35 is reduntant, since constexpr is already const

Comment (makortel) [reply hls-fix.patch]: Converting to `static constexpr` is correct, and @iarspider is also right that the `const` is redundant here.

File: hls-fix.patch (hunk 2 above)
Comment (makortel) [reply hls-fix.patch:48]: Looks good, I'd keep the comments though (even if they don't add much information).

**Comment by smuzaffar** [comment]
please test workflow 29634.702

**Comment by smuzaffar** [comment]
+externals

**Pull request merged**